src/portable/link.rs

```diff
@@ -198,32 +213,43 @@ pub fn link(cmd: &Link, opts: &Options) -> anyhow::Result<()> {
         }
     );
     let mut config = config.with_cert_verifier(verifier.clone());
-    let connect_result = connect(&config);
-    if let Err(e) = connect_result {
-        if e.is::<PasswordRequired>() {
-            let password;
-
-            if opts.conn_options.password_from_stdin {
-                password = tty_password::read_stdin()?
-            } else if !cmd.non_interactive {
-                password = tty_password::read(format!(
+    let mut conn = connect(&config).await;
+
+    if let Err(e) = conn {
+        if let Some(err) = e.downcast_ref::<edgedb_tokio::Error>() {
+            if err.is::<PasswordRequired>() {
+                let password;
+
+                if opts.conn_options.password_from_stdin {
+                    password = tty_password::read_stdin()?
+                } else if !cmd.non_interactive {
+                    password = tty_password::read(format!(
                         "Password for '{}': ",
                         config.user().escape_default()))?;
-            } else {
-                return Err(e.into());
-            }
+                } else {
+                    return Err(e.into());
+                }
+
+                config = config.with_password(&password);
+                creds.password = Some(password);
+                if let Some(cert) = &*verifier.cert_out.lock().unwrap() {
+                    let pem = pem::encode(&pem::Pem::new("CERTIFICATE", cert.to_vec()));
+                    config = config.with_pem_certificates(&pem)?;
+                }
 
-            config = config.with_password(&password);
-            creds.password = Some(password);
-            if let Some(cert) = &*verifier.cert_out.lock().unwrap() {
-                let pem = pem::encode(&pem::Pem::new("CERTIFICATE", cert.to_vec()));
-                config = config.with_pem_certificates(&pem)?;
+                let mut conn: Connection = connect(&config).await?;
+                config = ask_branch(&mut conn, &mut creds, config).await?
+            }
+            else {
+                return Err(e.into());
             }
-            connect(&config)?;
         } else {
             return Err(e.into());
         }
+    } else if let Ok(mut conn) = conn {
+        config = ask_branch(&mut conn, &mut creds, config).await?
     }
```

Comment:
Here, you have:

```rust
if let Err(...) = x {
   ...
} else if let Ok(...) = x {
   ...
}
```

The idiomatic way to express this pattern would use `match`:

```rust
match x {
    Err(...) => ...
    Ok(...) => ...
}
```

(I think rust-analizer even has a "quick action" for this)